internal/driver/driver.go

```diff
@@ -122,6 +140,95 @@ func (d *Driver) createClientOptions() ([]opcua.Option, error) {
 	return opts, nil
 }
 
+// Gets the username and password credentials from the configuration.
+func (d *Driver) getCredentials(secretPath string) (config.Credentials, error) {
+	credentials := config.Credentials{}
+	timer := startup.NewTimer(d.serviceConfig.OPCUAServer.CredentialsRetryTime, d.serviceConfig.OPCUAServer.CredentialsRetryWait)
+	service := service.RunningService()
+	var secretData map[string]string
+	var err error
+	for timer.HasNotElapsed() {
+		secretData, err = service.SecretProvider.GetSecret(secretPath, secret.UsernameKey, secret.PasswordKey)
+		if err == nil {
+			break
+		}
+
+		d.Logger.Warnf(
+			"Unable to retrieve OPCUA credentials from SecretProvider at path '%s': %s. Retrying for %s",
+			secretPath,
+			err.Error(),
+			timer.RemainingAsString())
+		timer.SleepForInterval()
+	}
+
+	if err != nil {
+		return credentials, err
+	}
+
+	credentials.Username = secretData[secret.UsernameKey]
+	credentials.Password = secretData[secret.PasswordKey]
+
+	return credentials, nil
+}
+
+func generateCert() (*tls.Certificate, error) {
+
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate private key: %s", err)
+	}
+
+	notBefore := time.Now()
+	notAfter := notBefore.Add(365 * 24 * time.Hour) // 1 year
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate serial number: %s", err)
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization: []string{"Test Client"},
+		},
+		NotBefore: notBefore,
+		NotAfter:  notAfter,
+
+		KeyUsage:              x509.KeyUsageContentCommitment | x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageDataEncipherment | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		BasicConstraintsValid: true,
+	}
+
+	host := "urn:testing:client"
+	if ip := net.ParseIP(host); ip != nil {
+		template.IPAddresses = append(template.IPAddresses, ip)
+	} else {
+		template.DNSNames = append(template.DNSNames, host)
+	}
+	if uri, err := url.Parse(host); err == nil {
+		template.URIs = append(template.URIs, uri)
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, publicKeys(priv), priv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create certificate: %s", err)
+	}
+
+	certBuf := bytes.NewBuffer(nil)
+	if err := pem.Encode(certBuf, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}); err != nil {
+		return nil, fmt.Errorf("failed to encode certificate: %s", err)
+	}
+
+	keyBuf := bytes.NewBuffer(nil)
+	if err := pem.Encode(keyBuf, pemBlockForKeys(priv)); err != nil {
+		return nil, fmt.Errorf("failed to encode key: %s", err)
+	}
+
+	cert, err := tls.X509KeyPair(certBuf.Bytes(), keyBuf.Bytes())
```

Comment:
cert should be a local variable that is returned as a result of the function.

Reply:
What do you mean? It is a local variable or am I missing something?